File: aitools/proofs/components.py
from enum import Enum, auto
from typing import Callable, Optional, List, Dict, Any

from aitools.logic import LogicObject, Substitution, Variable, LogicWrapper
from aitools.logic.utils import map_variables_by_name, VariableSource, normalize_variables


class HandlerArgumentMode(Enum):
    RAW = auto()
    MAP = auto()
    MAP_UNWRAPPED = auto()
    MAP_UNWRAPPED_REQUIRED = auto()
    MAP_UNWRAPPED_NO_VARIABLES = auto()
    MAP_NO_VARIABLES = auto()


class HandlerSafety(Enum):
    TOTALLY_UNSAFE = auto()
    SAFE_FOR_HYPOTHESES = auto()
    SAFE = auto()


# TODO make this generic in the handler return type
class Component:
    def __init__(self, *,
                 listened_formula: LogicObject, handler: Callable,
                 argument_mode: HandlerArgumentMode, pass_substitution_as=...,
                 pure: bool, safety: HandlerSafety):

        self._normalization_variable_source = VariableSource()

        self.handler = handler
        listened_formula, _ = normalize_variables(listened_formula)
        self.listened_formula = listened_formula
        self.argument_mode = argument_mode
        self.pure = pure
        self.safety = safety

        self._func_arg_names = handler.__code__.co_varnames[:handler.__code__.co_argcount]

        pass_substitution_as = self.__validate_pass_substitution_as(pass_substitution_as)

        self.pass_substitution_as: Optional[str] = pass_substitution_as

        self.variables_by_name = (
            None if self.argument_mode == HandlerArgumentMode.RAW
            else map_variables_by_name(listened_formula)
        )

        self.__validate_handler_arguments(pass_substitution_as)

    def __validate_pass_substitution_as(self, pass_substitution_as):
        if self.argument_mode == HandlerArgumentMode.RAW:
            if pass_substitution_as is None:
                raise ValueError(f"A substitution MUST be passed with {HandlerArgumentMode.RAW}")
            elif pass_substitution_as is Ellipsis:
                pass_substitution_as = 'substitution'

        else:
            if pass_substitution_as is Ellipsis:
                pass_substitution_as = None
        if isinstance(pass_substitution_as, str) and not pass_substitution_as.isidentifier():
            raise ValueError("When 'pass_substitution_as' is a string it must be a valid python identifier")
        return pass_substitution_as

    def __validate_handler_arguments(self, pass_substitution_as):
        if self.handler.__code__.co_posonlyargcount > 0:
            # TODO allow also kw-only args
            raise ValueError("Handlers cannot have positional-only arguments")

        if self.argument_mode == HandlerArgumentMode.RAW:
            if self._func_arg_names != ('formula', self.pass_substitution_as):
                raise ValueError(f"The handler has the wrong argument names {self._func_arg_names}! "
                                 f"{HandlerArgumentMode.RAW} requires the handler to take two arguments: "
                                 f"'formula' and '{pass_substitution_as}' (from the 'pass_substitution_as' arg)")
        else:
            unlistened_arg_names = list(
                arg_name
                for arg_name in self._func_arg_names
                if arg_name not in self.variables_by_name and arg_name != self.pass_substitution_as
            )
            if any(unlistened_arg_names):
                raise ValueError(f"The handler has the wrong argument names {self._func_arg_names}! "
                                 f"Handler arguments {unlistened_arg_names} "
                                 f"are not present in formula {self.listened_formula}")

    def _map_substitution_to_function_args(self, substitution: Substitution, func_arg_names: List[str],
                                           normalization_mapping: Dict[Variable, Variable]) -> Dict[str, Any] :
        prepared_args = {}
        for arg in func_arg_names:
            if arg in self.variables_by_name:
                mapped_variable = normalization_mapping.get(self.variables_by_name[arg], None)
                if mapped_variable is None:
                    raise ValueError("What is happening here?")
                prepared_args[arg] = substitution.get_bound_object_for(mapped_variable)
        return prepared_args

    def _extract_args_by_name(self, formula: LogicObject, unifier: Substitution,
                              normalization_mapping: Dict[Variable, Variable]) -> Dict[str, Any]:
        if self.argument_mode == HandlerArgumentMode.RAW:
            if self.pass_substitution_as is None:
                raise ValueError("NOOOOOOOOOO! WHAT HAVE YOU DONE???????")

            args_by_name = {'formula': formula}
        else:
            args_by_name = self._map_substitution_to_function_args(substitution=unifier,
                                                                   func_arg_names=self._func_arg_names,
                                                                   normalization_mapping=normalization_mapping)

            if self.argument_mode == HandlerArgumentMode.MAP:
                pass
            elif self.argument_mode == HandlerArgumentMode.MAP_NO_VARIABLES:
                if any(isinstance(arg, Variable) for arg in args_by_name.values()):
                    raise ValueError()
            elif self.argument_mode == HandlerArgumentMode.MAP_UNWRAPPED:
                args_by_name = {
                    key: arg.value if isinstance(arg, LogicWrapper) else arg
                    for key, arg in args_by_name.items()
                }
            elif self.argument_mode == HandlerArgumentMode.MAP_UNWRAPPED_REQUIRED:
                if any(not isinstance(arg, LogicWrapper) for arg in args_by_name.values()):
                    raise ValueError()

                args_by_name = {
                    key: arg.value if isinstance(arg, LogicWrapper) else arg
                    for key, arg in args_by_name.items()
                }
            elif self.argument_mode == HandlerArgumentMode.MAP_UNWRAPPED_NO_VARIABLES:
                if any(isinstance(arg, Variable) for arg in args_by_name.values()):
                    raise ValueError()

                args_by_name = {
                    key: arg.value if isinstance(arg, LogicWrapper) else arg
                    for key, arg in args_by_name.items()
                }
            else:
                raise NotImplementedError(f"Unsupported argument mode: {self.argument_mode}")

        if self.pass_substitution_as is not None:
            args_by_name[self.pass_substitution_as] = unifier

        return args_by_name

    def __str__(self):
        return f"{self.__class__.__name__}({self.handler.__module__}.{self.handler.__name__})"

File: AgainIsThisAGoodIdea/Proofs/Components.fs
module AITools.Proofs.Components.Base

open AITools.Logic.Language
open AITools.Logic.Utils
open AITools.Utils.Logger
open FSharp.Reflection
open AITools.Logic.Unification
open AITools.Logic.Core
open System

exception InvalidHandlerArgumentTypeException of string

type HandlerSafety =
    | TotallyUnsafe = 0
    | Safe = 1

type HandlerPurity =
    | Pure = 0
    | Impure = 1

type HandlerSubstitutionArgument =
    | PassSubstitutionAs of string
    | NoSubstitution

type HandlerContextArgument =
    | PassContextAs of string
    | NoContext

type RecordHandler<'handler> =
    { HandlerFunction: 'handler
      HandlerArguments: string array }

type BaseHandlerDescriptor<'handler> =
    { Handler: 'handler
      Expression: Expression
      Purity: HandlerPurity
      Safety: HandlerSafety }

type RawHandlerExtraArguments =
    { PassSubstitutionAs: HandlerSubstitutionArgument
      PassContextAs: HandlerContextArgument }

type MappedHandlerExtraArguments =
    { PassSubstitutionAs: HandlerSubstitutionArgument
      PassContextAs: HandlerContextArgument }

let makeMappedHandlerDescriptor ctor (expression, handler, passSubstitutionAs, passContextAs, purity, safety) =
    ctor
    <| ({ Handler = handler
          Expression = expression
          Purity = purity
          Safety = safety },
        { PassSubstitutionAs = passSubstitutionAs
          PassContextAs = passContextAs })
// TODO clean this, like... let's make it readable
type HandlerDescriptor<'handler> =
    | Raw of BaseHandlerDescriptor<'handler> * RawHandlerExtraArguments
    | Map of BaseHandlerDescriptor<'handler> * MappedHandlerExtraArguments
    | MapUnwrapped of BaseHandlerDescriptor<'handler> * MappedHandlerExtraArguments
    | MapUnwrappedRequired of BaseHandlerDescriptor<'handler> * MappedHandlerExtraArguments
    | MapUnwrappedNoVariables of BaseHandlerDescriptor<'handler> * MappedHandlerExtraArguments
    | MapNoVariables of BaseHandlerDescriptor<'handler> * MappedHandlerExtraArguments

    // TODO these are fun and all but I don't really like how they came out
    static member MakeRaw(expression, handler: 'handler, purity, safety, passSubstitutionAs, passContextAs) =
        Raw
        <| ({ Handler = handler
              Expression = expression
              Purity = purity
              Safety = safety },
            { PassSubstitutionAs = passSubstitutionAs
              PassContextAs = passContextAs })

    static member MakeMap(expression, handler: 'handler, purity, safety, passSubstitutionAs, passContextAs) =
        makeMappedHandlerDescriptor Map (expression, handler, passSubstitutionAs, passContextAs, purity, safety)

    static member MakeMapUnwrapped(expression,
                                   handler: 'handler,
                                   purity,
                                   safety,
                                   passSubstitutionAs,
                                   passContextAs) =
        makeMappedHandlerDescriptor
            MapUnwrapped
            (expression, handler, passSubstitutionAs, passContextAs, purity, safety)

    static member MakeMapUnwrappedRequired(expression,
                                           handler: 'handler,
                                           purity,
                                           safety,
                                           passSubstitutionAs,
                                           passContextAs) =
        makeMappedHandlerDescriptor
            MapUnwrappedRequired
            (expression, handler, passSubstitutionAs, passContextAs, purity, safety)

    static member MakeMapUnwrappedNoVariables(expression,
                                              handler: 'handler,
                                              purity,
                                              safety,
                                              passSubstitutionAs,
                                              passContextAs) =
        makeMappedHandlerDescriptor
            MapUnwrappedNoVariables
            (expression, handler, passSubstitutionAs, passContextAs, purity, safety)

    static member MakeMapNoVariables(expression,
                                     handler: 'handler,
                                     purity,
                                     safety,
                                     passSubstitutionAs,
                                     passContextAs) =
        makeMappedHandlerDescriptor
            MapNoVariables
            (expression, handler, passSubstitutionAs, passContextAs, purity, safety)


type ArgumentExtractor<'context> =
    Expression * Substitution * 'context * Map<Variable, Variable> * option<Map<string, Variable>> * seq<string> * HandlerSubstitutionArgument * HandlerContextArgument -> Map<string, obj>

type Component<'handler, 'context> =
    { ExtractArgsByName: ArgumentExtractor<'context>
      Language: Language
      ListenedExpression: Expression
      Handler: 'handler
      InputArgs: string seq
      VariablesByName: Map<string, Variable> option
      PassSubstitutionAs: HandlerSubstitutionArgument
      PassContextAs: HandlerContextArgument
      Purity: HandlerPurity
      Safety: HandlerSafety }

    override x.ToString() = x.ListenedExpression.ToString()

/// Generates a map-based handler and the list of its arguments from a record-based handler
let makeRecordHandler (handlerFunc: 'input -> 'output) =
    let inputType = typeof<'input>
    if not (FSharpType.IsRecord(inputType))
    then invalidArg "handlerFunc" "Only record types are allowed"

    let inputArgs =
        FSharpType.GetRecordFields(inputType)
        |> Array.map (fun f -> f.Name)

    let constructor =
        FSharpValue.PreComputeRecordConstructor(inputType)

    // helper to convert a Map into the 'input Record the handler accepts
    let handler func (data: Map<string, obj>) =
        let input =
            constructor (Array.map (fun f -> data.[f]) inputArgs) :?> 'input

        input |> func

    // a new handler function, which now accepts a Map and then calls the original handlerFunc
    let convertedFunc input = 
        try
            handler handlerFunc input
        with
        | :? InvalidCastException -> raise <| InvalidHandlerArgumentTypeException "Invalid argument types for record"

    { HandlerFunction = convertedFunc
      HandlerArguments = inputArgs }

// TODO this was a nice dream, but the type system didn't agree
// /// Generates a map-based handler and the list of its arguments from a record-based handler
// let makeFunctionHandler (handlerFunc:#obj -> #obj) =
//     let inputType = handlerFunc.GetType()
//     // TODO this is not really solid, is it? :P
//     let invoke = inputType.GetMethods() |> Array.find (fun m -> m.Name = "Invoke")
//     let inputArgs = invoke.GetParameters() |> Array.map (fun p -> p.Name)
//     let inputArgsList = inputArgs |> List.ofArray
//     //let boxF f (a: obj) = f <| downcast a
//     let convertedFunc (data: Map<string, obj>) =
//         let rec callWithArgs (f: obj) args = 
//             match args with
//             | [] -> f
//             | head :: tail -> 
//                 callWithArgs (downcast f <| data.[head]) tail
//         try
//             callWithArgs handlerFunc inputArgsList
//         with
//         | :? InvalidCastException -> raise <| InvalidHandlerArgumentTypeException "Invalid argument types for record"
//     { HandlerFunction = convertedFunc
//       HandlerArguments = inputArgs }


let prepareVariablesByName (expression, passSubstitutionAs, passContextAs) =
    let result = (mapVariablesByName expression)
    match passSubstitutionAs, passContextAs with
    | PassSubstitutionAs s, _ when result.ContainsKey s ->
        failwithf
            "Argument passSubstitutionAs conflicts with variable name \"%s\", they cannot be equal"
            s
    | _, PassContextAs s when result.ContainsKey s ->
        failwithf
            "Argument passContextAs conflicts with variable name \"%s\", they cannot be equal"
            s
    | _ -> result

let validateRawArgumentNames (inputArgs, passSubstitutionAs, passContextAs) =
    // the handler must accept an expression, a substitution and possibly a kb
    let inputSet = set inputArgs

    match passSubstitutionAs with 
    | NoSubstitution -> failwith "You need to specify a name for the substitution, but you didn't!"
    | PassSubstitutionAs subst -> 
        if not
            (
                Array.contains inputSet.Count [| 2; 3 |]
                && inputSet.Contains "expression"
                && inputSet.Contains subst
                && match passContextAs with
                    | PassContextAs s -> inputSet.Contains(s)
                    | NoContext -> inputSet.Count = 2
            )
        then
            let expected = 
                match passContextAs with
                | PassContextAs ctx -> ["expression"; subst; ctx]
                | NoContext -> ["expression"; subst]
            failwithf "The handler has the wrong argument names %A! expected %A" inputArgs expected


let validateMappedArgumentNames (inputArgs, variablesByName, passSubstitutionAs, passContextAs) =
    let unlistenedArgNames =
        Array.filter 
            (fun name ->
                not (Map.containsKey name variablesByName)
                && not
                    (PassSubstitutionAs name = passSubstitutionAs
                    || PassContextAs name = passContextAs)
            ) 
            inputArgs

    if not (Array.isEmpty unlistenedArgNames) then
        failwithf
            "The handler has the wrong argument names %A! Handler arguments %A are not present in the arguments"
            inputArgs
            unlistenedArgNames


// TODO this can be curried or somehow refactored
let mapSubstitutionToFunctionArgs (variablesByName: Map<_, _> option,
                                   substitution: Substitution,
                                   funcArgNames,
                                   normalizationMapping) =
    debug <| sprintf "mapSubstitutionToFunctionArgs"
    let mutable preparedArgs = Map.empty
    for arg in funcArgNames do
        debug <| sprintf "mapSubstitutionToFunctionArgs arg %s" arg
        if variablesByName.IsSome
           && Map.containsKey arg variablesByName.Value then
            match Map.tryFind (Map.find arg variablesByName.Value) normalizationMapping with
            | None -> 
                debug <| sprintf "mapSubstitutionToFunctionArgs what is happening?\n\tvariablesByName: %O\n\tfuncArgNames: %A\n\tnormalizationMapping: %O" variablesByName funcArgNames normalizationMapping
                failwith "What is happening here?"
            | Some mappedVariable ->
                match substitution.GetBoundObjectFor(mappedVariable) with
                | None -> failwithf "Variable %A is not mapped in substitution %A" mappedVariable substitution
                | Some boundObject -> preparedArgs <- Map.add arg (boundObject) preparedArgs

    preparedArgs

// TODO refactor arguments
let extractArgsByName argumentMode
                      (expression,
                       unifier,
                       context,
                       normalizationMapping,
                       variablesByName,
                       handlerArgs,
                       passSubstitutionAs,
                       passContextAs)
                      =
    debug <| sprintf "extraArgsByName"
    let argsByName =
        lazy (mapSubstitutionToFunctionArgs (variablesByName, unifier, handlerArgs, normalizationMapping))

    debug <| sprintf "extraArgsByName value = %O" argsByName.Value
    let baseArgs =
        match argumentMode with
        | Raw _ -> Map.add "expression" expression Map.empty
        | Map _ -> Map.map (fun _ v -> v :> obj) argsByName.Value
        | MapNoVariables _ ->
            if Map.exists (fun _ v ->
                match v with
                | Var _ -> true
                | _ -> false) argsByName.Value then
                raise <| InvalidHandlerArgumentTypeException "Unexpected variable!"
            else
                Map.map (fun _ v -> v :> obj) argsByName.Value
        | MapUnwrapped _ ->
            Map.map (fun _ v ->
                match v with
                | Wrap o -> o
                | _ -> v :> obj) argsByName.Value
        | MapUnwrappedRequired _ ->
            Map.map (fun _ v ->
                match v with
                | Wrap o -> o
                | _ -> raise <| InvalidHandlerArgumentTypeException "Unexpected unwrapped object") argsByName.Value
        | MapUnwrappedNoVariables _ ->
            Map.map (fun _ v ->
                match v with
                | Var _ -> raise <| InvalidHandlerArgumentTypeException "Unexpected variable!"
                | Wrap o -> o
                | _ -> v :> obj) argsByName.Value


    // TODO this is a job for a maybe monad but I'm too lazy
    let argsWithMaybeSubstitution =
        match passSubstitutionAs with
        | NoSubstitution -> baseArgs
        | PassSubstitutionAs name -> Map.add name (unifier :> obj) baseArgs

    let argsWithMaybeContext =
        match passContextAs with
        | NoContext -> argsWithMaybeSubstitution
        | PassContextAs name -> Map.add name (context :> obj) argsWithMaybeSubstitution

    argsWithMaybeContext


let prepareHandler (wrapHandler: 'a -> RecordHandler<'b>) handlerArgs =
    let lang = Language()

    match handlerArgs with
    | Raw (args, extraArgs) ->
        let listenedExpression, _ = renewVariables lang args.Expression
        let preparedHandler = wrapHandler args.Handler

        validateRawArgumentNames (preparedHandler.HandlerArguments, extraArgs.PassSubstitutionAs, extraArgs.PassContextAs)

        { ExtractArgsByName = extractArgsByName handlerArgs
          Language = lang
          ListenedExpression = listenedExpression
          Handler = preparedHandler.HandlerFunction
          InputArgs = preparedHandler.HandlerArguments
          VariablesByName = None
          PassSubstitutionAs = extraArgs.PassSubstitutionAs
          PassContextAs = extraArgs.PassContextAs
          Purity = args.Purity
          Safety = args.Safety }
    | Map (args, extraArgs)
    | MapUnwrapped (args, extraArgs)
    | MapUnwrappedRequired (args, extraArgs)
    | MapUnwrappedNoVariables (args, extraArgs)
    | MapNoVariables (args, extraArgs) ->
        let listenedExpression, _ = renewVariables lang args.Expression
        let preparedHandler = wrapHandler args.Handler

        let variablesByName =
            prepareVariablesByName (listenedExpression, extraArgs.PassSubstitutionAs, extraArgs.PassContextAs)

        debug <| sprintf "prepared variablesByName: %O" variablesByName

        validateMappedArgumentNames
            (preparedHandler.HandlerArguments,
             variablesByName,
             extraArgs.PassSubstitutionAs,
             extraArgs.PassContextAs)

        { ExtractArgsByName = extractArgsByName handlerArgs
          Language = lang
          ListenedExpression = listenedExpression
          Handler = preparedHandler.HandlerFunction
          InputArgs = preparedHandler.HandlerArguments
          VariablesByName = Some variablesByName
          PassSubstitutionAs = extraArgs.PassSubstitutionAs
          PassContextAs = extraArgs.PassContextAs
          Purity = args.Purity
          Safety = args.Safety }
